### ISSUE TITLE
Repository branching and tagging, versioning

### ISSUE THREAD

**Issue by ThomDietrich** [homieiot/convention#46]
Hello @marvinroger,

Somewhere in comments we already discussed that branching in this repository should be addressed. I want to restart that discussion. Here's my first suggestion, which is open for discussion:

1. **Introduce a strict versioning system** - Homie is a convention to be followed by other development projects. It is hence important to add unambiguous version numbers for every convention revision. A developer needs to be able to state "The *app* conforms with the Homie convention v2.1.3". I suggest to implement semantic versioning: http://semver.org (should be a perfect fit)

2. **Work in the `main` branch only** - Currently there is an outdated master branch, an unclear v2.0 branch and a redesign branch. That is neither intuitive nor visible to a reader. As the convention is steadily evolving, we should present the latest revision to the interested developer/user. (With the option to switch to other versions, see 5)

3. **Use tags to mark versions** - A homie version is a fixed state of the document, hence tags are perfectly fine for the job. (Other [branching models](http://nvie.com/posts/a-successful-git-branching-model/) suggest branches in order to allow e.g. hotfixes)

4. **Version every commit** - Almost every commit will change details of the Homie convention, it is therefore needed to increment the version number as the MAJOR.MINOR.PATCH system of semver suggests. (Exceptions are of course commits with language changes only)

4. **GitHub releases for stable tags** - GitHub offers the [releases tab](https://github.com/marvinroger/homie/releases) to publish versions and add a release note. That should be used to announce and document the stable and noteworthy releases (e.g. v1.5.0, v2.0.0, v2.1.0)

5. **Link the "Tags" tab in the first part of the convention** - For readers to quickly find all available homie versions, the [tags tab](https://github.com/marvinroger/homie/tags) should be highlighted in the first part of the convention document.

Everyone, please add your comments, criticism and feedback!
Best! Thomas 🍻

**Comment by marvinroger** [comment]
Hi @ThomDietrich 
Thanks for that! Actually, this was already planned (we do exactly all of that on https://github.com/marvinroger/homie-esp8266). I wanted to "clean everything" as soon as we all agreed about the v2.1.0, which will be the last release that breaks semver (the v2.0.0 is not really backward compatible, but I'd say this is fine as the v2.0.0 was never released officially).

The only thing I did not think about was the "Tags" tab. Great idea! 👍

**Comment by marvinroger** [comment]
https://github.com/marvinroger/homie/commit/e0e3bad3c659c12146819d53b2612098acedc176 ✨

**Comment by ThomDietrich** [comment]
@marvinroger just a thought before it's too late. After all recent additions and changes I wonder if 2.1.0 is the correct version to choose. How about 3.0.0 or 2.5.0?

**Comment by marvinroger** [comment]
I agree that v3.0.0 would be semver-compliant. I don't know, because the v2 was actually never released. But yes, the versioning was chaotic. Let's follow semver and go for v3. 😈

We'll have it ready very soon, that'll be a thing of the past. 😉

**Comment by bodiroga** [comment]
Hi @marvinroger!

What's left for the release of the official Homie 3.0 version? All PR's are merged and all big issues are closed, what else needs to be done?

Many thanks for all your hard work! :+1:

**Comment by ThomDietrich** [comment]
@marvinroger I really think the current repo situation is all other than ideal. We should do something about that... You can always start by creating a beta release or similar... Everything is better than the current state.

**Comment by timpur** [comment]
Added some tags and releases, getting close :)

**Comment by ThomDietrich** [comment]
@timpur thanks for pushing development forward. I'd suggest to do a few things right away, in order to finally reach a production ready and extendable state:

* change the version of the redesign branch as it is right now to v3.0. (This version step was discussed above and your ideas over in #61 can just as well be added with 3.1, 3.5 or 4.0.)

* Merge redesign into master, delete all other branches

* Add and clean up tags

... So basically enforce everything I've noted down in the first message here 😃 What do you think?

**Comment by ThomDietrich** [comment]
@timpur did you miss the mentioning?

What do you think? If you are not disinclined, I'd go and apply the plan tomorrow.

**Comment by timpur** [comment]
Yes, just stretched for time atm ....

**Comment by ThomDietrich** [comment]
@timpur @marvinroger 
I might have overestimated my free time back in the last comment :D
Just so we are all on track: Do we want to go forth with the plan to mark version 3.0 now, then address all existing issues? A simple 👍 would be enough to show your approval. 😉 I'll do the needed work then.

**Comment by timpur** [comment]
What are some of the big issues atm, tbh not following them all ... but yest i think were ready for 3.0

**Comment by marvinroger** [comment]
Definitely!

Sorry for my tiny presence here, I have been preparing my internship
defense, which is tomorrow! 😊

Le lun. 19 févr. 2018 à 13:56, Thomas Dietrich <notifications@github.com> a
écrit :

**Comment by ThomDietrich** [comment]
@marvinroger Good luck!

**Comment by timpur** [comment]
@ThomDietrich, Do what you can and ill help when i can :)

Thank for contributing.

**Comment by bodiroga** [comment]
Hope that everything went fine @marvinroger!

Thanks for all your work guys!

**Comment by timpur** [comment]
Question, how do you guys prefer to do implementation versioning? eg ESP implementation? Should it align with Homie version?

**Comment by ingoogni** [comment]
Yes, IMHO it should be clear what versions of both match

**Comment by timpur** [comment]
yes but what :P ?

**Comment by ingoogni** [comment]
H3.0ESP23.04

**Comment by timpur** [comment]
Damn

**Comment by timpur** [comment]
Could be:
c = Homie Convention
i = Homie Implmentaton version
`c.c.i.i`
so for Homie ESP currently that would look like `2.0.2.0` think @nerdfirefighter said this.
but not sure about it, i think they should be separate, just need to document these version better, and display them more !!!

**Comment by timpur** [comment]
@marvinroger need you to remove v2 branch, and set master as default. I have merged and cleaned up branch v2 so you can remove.

Master sits at v2.0 now and is ready for redesign branch (3.0) we will work on redesign and master will reflect most recent release always.

We now need to get 3.0 ready for release.

**Comment by ThomDietrich** [comment]
@timpur I have to apologize. I ran into some unforeseen problems in my personal endeavors and didn't get around to continue here. Thanks for taking matters into your hands.

One thought I want to mention: You've changed "2.1.0" to "3.0" but it should read "3.0.0". Is this a mistake? Please compare: http://semver.org/

**Comment by timpur** [comment]
Didn't know what existed .... Good to know. My mistake, will fix

**Comment by marvinroger** [comment]
Hi guys!

Thanks for the work being done here. I did what you've asked @timpur. I just made some cleanup to my notifications (had a few hundreds), so I hope I'll be able to be more responsive.

**Comment by timpur** [comment]
thanks @marvinroger, sorry to bombard you ...

**Comment by marvinroger** [comment]
Don't be sorry! 😇

**Comment by ThomDietrich** [comment]
Marvin: Thanks!

@timpur I'm confused about the current state. The idea was to merge redesign into master. Are you going to do so? See point 2 of my list in OP...

**Comment by timpur** [comment]
IMHO, I think master should reflect the latest release and we should be working from a device branch... This is how most repos work so why change it here ?

**Comment by ThomDietrich** [comment]
I thought to have argued about that in the first post.
The important thing is to remember that the Homie Convention is destined to be revised constantly. That's why Marvin and I agreed to go with a stringent version scheme. Every addition or change to the convention will potentially be picked up by developers interested in this new aspect of the convention. Hence a version for this addition/change is needed.
To cope with those requirements we agreed to develop in master and pull requests only. To make sure every developer is still able and encouraged to access specific versions, we've introduced this link:

![grafik](https://user-images.githubusercontent.com/2870104/37590698-5f3223ea-2b69-11e8-955b-669349a68bd1.png)

**Comment by timpur** [comment]
Im totally fine with that route. But i just have one question. 

What if were half way in releasing a new version and working from master and a new dev comes in and starts to use what they see in master assuming that that is the latest release (as most repos use that method) and then later realises what they have implemented is half baked.... 

Will we need to make sure people only use versions from the release tab? Like we should only be using versions from releases but you know....

I just wonder what is the benefits and what do you risk.

**Comment by ThomDietrich** [comment]
A good thought. Let's presume that the master branch only evolves through PRs. Half baked changes shouldn't find their way into the master branch because they are fine-tuned in the PR first. The goal is to have quick and direct iterations of the outwards-facing convention. Every non-trivial PR merge will be accompanied by a version increment in the form of a tag.

Now how is a developer of an application implementing the convention supposed to select a Homie Version he/she wants to conform to? (This is related to #51 )
I'd say first and foremost they should always refer to the latest revision if possible. I'm saying "if possible" because it really depends on the *other side*, the controller or node the dev wants to interact with. He/she will need to select a specific version if in doubt. In this moment a conservative master branch wouldn't do much good as we can't predict which Homie convention version the *other side* is working with. In the end the developer needs to compare versions (semver already addresses compatibility) and the changelog in order to select the right version to implement.

Looking at the above I can't make out any advantages of a dedicated develop branch, on the other side I am seeing additional delays and overlaps growing out of it.

I'm not trying to undermine your critique. I may be utterly wrong and be missing something :) Let me know if you are still not convinced.

**Comment by timpur** [comment]
TBH, same outcome so i dont mind, if you think it will work then i trust you. If it doesnt we can always go back to master dev... so lets give it a shot. I like what your trying to achieve.

**Comment by euphi** [comment]
I agree with @ThomDietrich.
However, never say never. So if one day there is a need for a larger redesign or a new complex "feature" there should be a dev branch again.
The advantage of such a branch is that it is easier to go "backwards".
If something was added to master (and tagged) and later reverted because someone had a better idea, you get incompatible official versions.

**Comment by nerdfirefighter** [comment]
I'm no git/repo expert so this might be completely left field, but i'm going to throw it in the ring away.

if i understand the constraints and intentions correctly i see the following:
* once a major version is released, then changes to that version are mainly
  * bug fixes (particularly in the documentation of the version)
  * addition of non mandatory topics
    * things that make it easier to do something or standardise something but not something that dictates how to use the framework. e.g. defining an "standard" attribute of $unit for version 2.0 for a property so that we can make a standard connector in Home Assistant for it, but there is no need to use it if you dont want to.
  * other cosmetic changes. 

So as i sort of see already, we need/should be able to maintain older versions with out affecting the latest development or main version. so in this case, there are things that would be nice to to with V2 with out the need to change over all the code just to implement V3 when V3 is not in a released state any way. 

Now i see a few options

1. A new Repo per major version - This would get messy and people would not know which repo to use. - Not worth it.
2. A new branch pair(dev and master) per version. e.g. master-v2, develop-v2. this would allow independent development, bug fixes, etc with out  the need to worry about updates from other versions of the framework being released before they are ready or being affected if we need to rollback. its not much extra overhead as we dont create new major versions that often. also master would be more of a pointer branch that directed people to the branch of the version they want
3. Or, we restructure the layout and put each version in a top level folder, minor versions in a sub folder of that version and have two branches, dev (things that we are not in release state) and master (latest stable version). this has the same issue as what we have now and option 1 in promoting one version to master may push up changes in a different version of the framework while it is still in development and not in a stable state. 

just my two cents for thinking about.

**Comment by timpur** [comment]
@ThomDietrich #68 is an example of why we normally have master and dev(another branch), while we're working on 3.0 in dev additions can be made to the current release.... this is hard with one branch.

**Comment by euphi** [comment]
If there is a need to make changes to an old release, then you can create a branch from the old release tag. There is no need to have a branch before a change is needed.

**Comment by timpur** [comment]
@euphi What do you do with the branch you create from an older tag, you can merge that back into the master branch with newer changes..... sounds messy...

**Comment by euphi** [comment]
For #68:

Variant a) Do the change for v2.0.1 in master branch and tag it as 2.0.1. Later merge "redesign" into this.
Variant b) Create a v2.x-branch right now, do the change in that branch and tag it as 2.0.1. No need the merge "redesign" into this.

I prefer b).

**Comment by timpur** [comment]
My question is after tagged how do you get it back into master when master has 3.0.0 in it ... We don't want to leave branches hanging around the....

**Comment by ThomDietrich** [comment]
Interesting. In my whole concept proposal I did not incorporate the need to change later versions. If you need `$nodes` I'd say you should use a version that supports it. Why the extra effort to patch older versions? You'll in any case need to adapt the controller and the node to support `$nodes`. Why not just go with v3.0.0 then?

@euphi's proposal would be how I would do it... however I wouldn't do it 😄

**Comment by ThomDietrich** [comment]
Naaah don't say that, we are all equals here and decisions based on diverse needs and perspective are a great thing. As I said before, my opinion is as it is but I'm more than willing to find a compromise everyone* is happy with.

I'd say we can't compare the repository to any other, because it is not a software we are developing, it's a standard.

My fear with either dev and/or version branches is the confusion and work arising out of them. Looking back at the last halve year, neither @marvinroger nor I and others had much time to improve the convention in all directions that were discussed. Still we were able to address the most serious issues with regard to *Discoverability* - the most important requirement wrt the Homie convention.

To drive Homie forward it is my impression, that we need to clear out uncertainties and reach a level of definition and documentation of a latest version that is sufficient for users of all levels.

I am not against your proposal in #68, I would only give to reason that your contribution would have a higher return if we all worked towards a better version 3 to reach the above. Now once again, if you think #68 is meaningful to you *and* (!) others, feel free to shoot away ;)

**Comment by marvinroger** [comment]
Also guys, one thing: we talked with @timpur about moving the convention to https://github.com/homieiot

I can understand people might not be confortable with the idea of working with something (e.g. Homie) that's tied to an individual. What do you think?

**Comment by lorenwest** [comment]
Since you asked...

I've found nothing comparable to Homie in terms of convention and reference implementation. It's a great piece of work, and setting it free is a powerful statement from the original artist @marvinroger. It's a great statement about your commitment to open source, and I couldn't agree more with moving it to https://github.com/homieiot.

Now that I'm typing, I should add that as an author of a fairly successful collaborative [open source project](https://github.com/lorenwest/node-config), branches and version numbering are important for clients of the specification (and implementations).

The first impression github page should be of the stable code (or convention in this case). Master is for stable production use, and branches are for research & development. The signal that something is production grade is that it's merged into master. Of course that's just a convention, but the reasoning behind it is sound.

As for versioning, the [semver spec](https://semver.org/) is a great guide and arbitrator. It's a simple way to communicate complex concepts.

Thank you for the hard work that everyone's putting into this. I've been following updates and will update the [NodeJS implementation](https://github.com/microclimates/homie-device) as the specification stabilizes.

...back to lurking

**Comment by ThomDietrich** [comment]
@marvinroger as you might remember I proposed something similar in one of our mails. Hence fine with me ;)

**Comment by ThomDietrich** [comment]
One more try to settle a few comments regarding branches:

1. PRs are effectively branches. No unstable "code" will find it's way into master
2. The idea to constantly have a `dev` branch is founded on the believe, that there will be *one* stable version we should present on master - I am not convinced this is true.
3. Users will want to browse specific versions of the convention (not any single one). That's why the reference to the tags/releases tab and the previously proposed list of significant versions is important.

@lorenwest did this scatter your fear?

Concluding: I would proposed to make a clean cut now. Reduce to master, add tags for major versions, and continue with the improvement discussions. If at a later point bigger changes in the form of multiple additions are meaningful, I'll personally create a branch named `develop` 😃

**Comment by marvinroger** [comment]
I do like [GitHub Flow](https://guides.github.com/introduction/flow/). It's nice and easy to understand, and it's yet what has worked the best for me.

The idea is simple: `master` should contain the latest "stable" (in our case, the latest released convention). Whenever we want to make a new version, we create a new specific branch. This allows for small and localized increments. If the change only contain tiny fixes (typo, graphical changes), then it can be merged to `master` without bumping version. Otherwise, the version would need to be bumped prior to merging.

I am also against maintaining old versions, and having a `develop` branch. From my experience, having two parallel branches is a true nightmare to manage on GitHub. Just one extremely annoying thing: do a PR from `develop` into `master`. You'll have a nice commit history. Once merged, continue to work on the `develop` branch, and once again, do a PR against `master`. You'll not only have the newest commits, but also the ones prior to the first merge. So it gets worse on every merge. Plus, after each merge you have to rebase your `develop` to the tip of `master`. I've found this pattern hard to follow.

**Comment by lorenwest** [comment]
@ThomDietrich - Yes to everything.

In my opinion, branches are good for:

1) Collaborating on future major versions (transient - soon to be master)
2) Archiving past major versions (long lived)
3) Ad-hoc collaboration on an issue (transient - for the life of the feature)

1 being a subset of 3 means there's only transient issue branches and major release branches.

Having a single long lived `dev` branch doesn't have much value, and tagging important commits within a major release can be valuable to communicate a point in time.

**Comment by ThomDietrich** [comment]
@marvinroger not sure if we are on the same page here. The GitHub flow workflow is basically what I am proposing from the beginning 😃 you just have to remember that PRs automatically mean that there is a dedicated branch for the change.

**Comment by euphi** [comment]
Yeah, this is also something I wanted to write: Each PR is based on a branch. So the proposal of @ThomDietrich includes to have (short-living) branches for feature implementation.
For more complex features I already (yesterday) proposed to create the branch manually (but later merge it back to master).

Furthermore, for "patching" old releases, you **must** create a branch. So, for #68 , you need to create a branch the branches off at the commit with the v2.0.0 tag. We should agree on a convention how to name this branch, e.g. release-v2.x .

@marvinroger it seems that both [Github Flow](https://guides.github.com/introduction/flow/) and [Git Flow](https://gitversion.readthedocs.io/en/latest/git-branching-strategies/gitflow/) don't handle this "Number #68"-case that you want to apply "patches" to old revisions. 

So, my proposal (based on this discussions here):

1. Work on master branch
  Alternative: Create development branch and merge back to master only for releases
2. Releases **shall** be on master branch with Semver-Compliant Tag for each release.
  -> this is a specification/convention, so there **should** be no need for "beta" versions.
3. Fixes or compatible improvements to old versions **may** be created and if so they **shall** be on a branch that is named _release-v{majorversion}.{minorversion}.x_.. 
  --> There **shall** be no incompatible changes to old versions, so the {majorversion}.{minorversion}.x scheme is always sufficient to be semver-compliant.
  --> Fixes on these branches *may* be merged back to master (or develop) 


I hope we all agree with 2. and 3. - at least it seems that they are not affected by the decision regarding 1.

**Comment by timpur** [comment]
Translating what seems to be coming out of this is we should be creating a branch for a new feature or change. Up the version on that branch, create a PR for that branch to master. When approved, merge and delete branch, and create release?

**Comment by euphi** [comment]
@timpur yes, but github alllows you to directly create a PR. The necessary branch is created automatically and github asks to delete the branch, if the PR is merged. 

So, work in master branch, but always use PRs for changes (even if you have the rights to directly edit in the master branch) is exactly what you describe in your first paragraph.

Every new "feature" can be released immediatly, so as long as there no new features that are too large to handle it in one PR, there is no need for a development branch.

**Comment by timpur** [comment]
Makes sense. So for bigger changes that require multiple people make a feature branch... And remove after release back to master.

**Comment by ThomDietrich** [comment]
To summarize, the next steps would be:

1. I'll merge redesign into master, sorting out leftover mess.
2. I'll check and complete tagged versions (some intermediate are missing)
3. I'll propose a PR adding details regarding the different versions of the Homie convention as described in point 3 [here](https://github.com/marvinroger/homie/issues/46#issuecomment-375030795)
4. You @timpur can implement #68 
5. We can ping all stalled issues to restart discussions based on the v3.0.0 state of the convention.
6. We can finally continue working on the actual convention instead of discussing technical details.
7. When a time for bigger refactoring comes, we will use a dedicated feature branch for that.

**Comment by timpur** [comment]
I don't understand how I can do #68 when you've merged 3.0.0 into master. I require 2.0.0, which I can branch off yes but I'll never be able to get it back into master when you've merged 3.0.0 ?

**Comment by euphi** [comment]
Yes, but what's your problem with a v2.0.0 branch? There is no need to merge it back.
That's why branches were invented... :-)

**Comment by ThomDietrich** [comment]
Exactly. You will see that everything falls into place once you start working on it.
Maybe this way of thought helps: In the end not your branch is the important artifact, your tag is. You will branch off of the v2.0.0 tag, add your changes in one or multiple commits, then create a tag "v2.0.1".
This tag is the important thing and is what will be available to users later. At this point no one cares on which branch it is based.

@euphi regarding the branch name. After consideration, I wonder if we should prefer `patch-v2.0.x` over `release-v2.0.x`?

**Comment by timpur** [comment]
Okay, if you dont thing git will garbage collect the unmerged commits in a deleted branch with a tag then sure. Im all in.

**Comment by euphi** [comment]
@timpur 
Who wants to deleted the branch?
There is absolutely no reason to do this.

**Comment by euphi** [comment]
@ThomDietrich 
I wanted to avoid the "hotfix-v2.0.x" branch name. As the "release-v2.0.x" name is proposed in the "git flow"-branching model, but we don't follow this model*, I'm fine with another name.

* in that model, the release-branch is created for developing the release and then merged back into master for the release. The git flow model does not offer a solution how to maintain old releases. (Not even for hotfix?!).

**Comment by ThomDietrich** [comment]
@marvinroger the master branch is protected and I can't push destructive changes. (Good so far! 😉)
However it would be meaningful to remove the latest commit on master to end up with a clean commit graph. Could you please do the following? Thanks!

```
git checkout master
git pull
git reset --hard 3bcbb01ef31205a5c2a19522fc9b6dd892d1c274
git merge --ff-only origin/redesign
git push
```

**Comment by marvinroger** [comment]
@ThomDietrich done 😉

**Comment by timpur** [comment]
Wait I thought the whole point of this was to only have one branch master and use feature branches only to make changed but they eventually get merged and deleted ? So won't 2.0.1 be the same. I'll tag it and relase it and then delete the branch ?

**Comment by euphi** [comment]
No, that's just not possible.
If you want to maintain a old release, you must have a branch for it.

The github flow and git flow models don't consider maintaining an old release, so we can't use them.

* So develop in master or feature branches.
* Release (tag) on master
* create release branches only to develop and tag a maintenance version of an old release

**Comment by marvinroger** [comment]
Are we ready to move to @homieiot? No immediate concerns about that?

I am just wondering, should we name it homieiot/homie or homieiot/convention? One thing, I'd like to create a website that would render the convention markdown, and having it as https://homieiot.github.io/convention sounds cleaner than https://homieiot.github.io/homie

**Comment by marvinroger** [comment]
Oh and yes @ThomDietrich, I remember that we already discussed about moving the convention to its own org. I was then looking for a job, and my GitHub weighted more than my CV. I am now in a comfortable position, so now that my career is "bootstrapped", I have no problem moving the convention. 😉

**Comment by euphi** [comment]
I also prefer "convention".

**Comment by euphi** [comment]
Regarding moving to homieiot? Should we fork the repository to keep history or should we create a new Repo?

For creating a website, the directory structure will change, so maybe a new repo is better?

So before moving the contents, we should clarify:

- Which site generator shall be used?
- When and how do we move content (the actual convention?)

**Comment by ThomDietrich** [comment]
@euphi we don'T have to worry about history. Moving a repo just means it will be shifted into another namespace. Btw. GitHub will auto-redirect the current repo URL to the new one so that links and users will still find their way.

@marvinroger we are in total sync :D I was also thinking about https://homieiot.github.io/convention 😉

**Comment by marvinroger** [comment]
Let's all welcome the new `homieiot/convention` repo 😄

**Comment by ThomDietrich** [comment]
Wonderful. You might want to adapt the short description of the project to include "Homie".

I'd say the issues of this issue are settled. However let's leave it open as a reference.

**Comment by marvinroger** [comment]
Just did. Thanks to everyone! 👏

**Comment by lorenwest** [comment]
Just noticed the [NodeJS implementation](https://github.com/homieiot/convention/pull/42/files) merged on [Sept. 21 2017](https://github.com/homieiot/convention/commit/e5dcd7e1908d168ecb1234cc480ef562aa1905aa) didn't make it to the [Home Page v2.0](https://github.com/homieiot/convention/tree/release-v2.0.x) or the new [Implementations](https://github.com/homieiot/convention/blob/master/implementations.md) page.

Probably an artifact of repository movement - although if it were an accident, I wonder what other commits fell through the cracks...

**Comment by timpur** [comment]
@lorenwest Done :)

**Comment by lorenwest** [comment]
Would you mind also adding back to https://github.com/homieiot/convention/tree/release-v2.0.x ? 

That's the version it was developed against. I'll update as soon as 3.x moves from WIP to LIVE.

**Comment by timpur** [comment]
@lorenwest Done, sorry should of thought about. Also you might want to define a version and link to one of the releases of Homie. https://github.com/homieiot/convention/releases :)
